tests/instruments/hp/test_hp3478a.py

```diff
@@ -0,0 +1,143 @@
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+from pymeasure.test import expected_protocol
+from pymeasure.instruments.hp import HP3478A
+
+VALID_CAL_DATA = [
+        0, 0, 0, 0, 3, 0, 8, 2, 15, 4, 4, 0, 13, 11, 0, 0,
+        0, 0, 3, 3, 2, 15, 5, 3, 0, 14, 0, 0, 0, 0, 0, 0,
+        3, 2, 15, 4, 0, 0, 14, 7, 9, 9, 9, 9, 9, 7, 2, 0,
+        15, 3, 12, 10, 11, 9, 9, 9, 9, 9, 9, 2, 0, 14, 15, 14,
+        9, 12, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 15, 15, 9,
+        9, 8, 6, 0, 9, 2, 14, 14, 0, 12, 10, 12, 9, 9, 9, 9,
+        9, 5, 1, 12, 0, 5, 14, 10, 13, 9, 9, 9, 9, 9, 8, 1,
+        12, 1, 15, 1, 10, 12, 0, 0, 0, 0, 0, 0, 1, 12, 15, 3,
+        0, 14, 0, 9, 9, 9, 9, 9, 9, 1, 12, 13, 2, 14, 9, 15,
+        9, 9, 9, 9, 9, 9, 1, 12, 13, 4, 2, 10, 9, 9, 9, 9,
+        9, 9, 9, 1, 12, 14, 12, 13, 9, 5, 9, 9, 9, 9, 9, 9,
+        1, 12, 2, 1, 15, 10, 10, 0, 0, 0, 0, 4, 2, 3, 0, 0,
+        3, 12, 14, 7, 0, 0, 0, 0, 0, 4, 3, 0, 1, 12, 3, 14,
+        8, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 15, 15, 9, 9,
+        8, 6, 0, 9, 3, 14, 3, 1, 1, 12, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0, 0, 0, 15, 15, 0, 0, 0, 0, 0, 0, 0, 0
+    ]
+
+
+def convert_cal_data_to_cal_read_xfers(cal_data):
+    # Convert cal_data into 256 single byte 'W' read operations.
+    cal_read_xfers = []
+    for addr, value in enumerate(cal_data):
+        cal_read_xfers.append([bytes([ord('W'), addr]), bytes([value+64])])
+
+    return cal_read_xfers
+
+
+def convert_cal_data_to_cal_write_xfers(cal_data):
+    # Convert cal_data into 256 single byte 'X' write operations.
+    cal_write_xfers = []
+    for addr, value in enumerate(cal_data):
+        cal_write_xfers.append([bytes([ord('X'), addr, value]), None])
+
+    return cal_write_xfers
+
+# ============================================================
+# TESTS
+# ============================================================
+
+
+def test_calibration_enabled():
+    with expected_protocol(
+            HP3478A,
+            [
+                (b"B", b'\x00\x20\x00\x00\x00'),        # cal_enable bit is set
+                (b"B", b'\x00\x00\x00\x00\x00')
+            ],
+    ) as instr:
+        assert instr.calibration_enabled
+        assert not instr.calibration_enabled
+
+
+def test_calibration_data_getter():
+    cal_read_xfers = convert_cal_data_to_cal_read_xfers(VALID_CAL_DATA)
```

Comment:
You use this line several times, why don't you make it a global variable (it is static) or a fixture?

Reply:
VALID_CAL_DATA is the root calibration data that belongs in the instrument SRAM.

The 2 convert_* functions are functions that transform it into GPIB transfers. I think it makes sense to have those factored as individual functions. It's something that's useful to have if others want to reuse it at some point for their own code.

After that:
*  test_calibration_data_getter creates a cal_read_xfers array, 
*  test_calibration_data_setter_pass creates a valid_cal_write_xfers array from VALID_CAL_DATA
*  test_write_calibration_data creates invalid_cal_write_xfers from a corrupted VALID_CAL_DATA

So cases of anything being used multiple times.